marconi/test/Spec/Utxo.hs

```diff
@@ -22,73 +26,180 @@ import Cardano.Api qualified as C
+    Hedgehog.assert . not . null $ sameAddressEvents
+    let eventsAddress :: Maybe (StorableQuery Utxo.UtxoHandle)
+        eventsAddress =
+          headOption
+          . nub
+          . toListOf ( folded . Utxo.address )
+          . concatMap (Set.toList . Utxo.ueUtxos)
+          $ sameAddressEvents
+    Hedgehog.assert . isJust $ eventsAddress
+    fromJust eventsAddress ===  head addresses
+
+txAddressToUtxoAddressTest ::  Property
+txAddressToUtxoAddressTest = property $ do
     t@(C.Tx (C.TxBody C.TxBodyContent{C.txOuts}) _)  <- forAll $ CGen.genTx C.BabbageEra
     let (targetAddresses :: Maybe TargetAddresses ) = addressesFromTxOuts txOuts
     let (utxos :: [Utxo.Utxo]) = Utxo.getUtxos targetAddresses t
     case targetAddresses of
-        Nothing         ->  (length utxos) === (length txOuts)
+        Nothing         ->  length utxos === length txOuts
         Just targets    ->
             ( nub
-              . catMaybes
-              . fmap (\x -> addressAnyToShelley (x ^. Utxo.utxoAddress))
+              . mapMaybe (\x -> storableQueryToShelleyAddr (x ^. Utxo.address))
               $ utxos) === (nub . toList $ targets)
 
-genBlockNo :: Gen C.BlockNo
-genBlockNo = C.BlockNo <$> Gen.word64 Range.constantBounded
-
-genEvents :: Gen Utxo.UtxoEvent
-genEvents = do
-    slotNo <- CGen.genSlotNo
-    blockNo  <- genBlockNo
-    txs <- Gen.list (Range.linear 2 5)(CGen.genTx C.ShelleyEra)
-    pure . fromJust $ Utxo.getUtxoEvents Nothing slotNo blockNo txs
-
 utxoStorageTest :: Property
 utxoStorageTest = property $ do
-    depth <- forAll $ Gen.int (Range.linear 1  5) -- force DB writes
-    ndx <- liftIO $ Utxo.open ":memory:" (Utxo.Depth depth)
-    events <- forAll $ Gen.list (Range.linear 1 10) genEvents
-    let (rows :: [Utxo.UtxoRow]) = nub . concatMap Utxo.toRows $ events
-        (addresses :: [C.AddressAny]) = nub . fmap (\r -> r ^. Utxo.utxoRowUtxo . Utxo.utxoAddress ) $ rows
-    ix <- liftIO $ Ix.insertL (events) ndx
-    rowsFromStore <- liftIO $ (concat . catMaybes) <$> forM addresses (\a -> Utxo.queryPlusVolatile ix a  )
-    (null rowsFromStore) === False
-    let (tid :: Utxo.UtxoRow -> C.AddressAny) = (\r -> r ^. Utxo.utxoRowUtxo . Utxo.utxoAddress)
-    (all (`elem` (tid <$> rows)) (tid <$> rowsFromStore)) === True
+  events <- forAll $ Gen.list (Range.linear 2 7) genEvents
+  (storedEvents :: [StorableEvent Utxo.UtxoHandle]) <-
+    (liftIO . Utxo.open ":memory:") (Utxo.Depth  10) -- no DB writes
+     >>= liftIO . Storable.insertMany events
+     >>= liftIO . Storable.getEvents
+  Hedgehog.assert (equivalentLists storedEvents events)
 
 addressesFromTxOuts :: [C.TxOut C.CtxTx CurrentEra] -> Maybe TargetAddresses
-addressesFromTxOuts [(C.TxOut addressInEra _ _ _)]
+addressesFromTxOuts [C.TxOut addressInEra _ _ _]
     = nonEmpty
     . nub
-    . catMaybes
-    . (fmap (addressAnyToShelley . Utxo.toAddr) )
+    . mapMaybe (addressAnyToShelley . Utxo.toAddr)
```

Comment:
Given that `addressesFromTxOuts` returns a `Maybe (NonEmpty a)`, then why not just `[C.Address C.ShelleyAddr]`?

Reply:
The signature for theTargetAddress is NonEmpty

Reply:
Right, so why not a `pure <$> addressAnyToShelley (Utxo.toAddr addressInEra)` then? (I.e why convert it to a single element list first at all and then operate on that when you don't need to. The `nub` is entirely unneeded there too, as the list will ever contain 0 or 1 value)

But a broader question about this function:  it takes a list of `[TxOut]` but only returns a `Just TargetAddresses` when the argument list has exactly one element and otherwise returns `Nothing`, why is that?

Reply:
the function is poorly named, but the comment still applies.
You're correct as the nub is nor required.
The main idea here is the create/make a TargetAddress that is relevant. Another words, a targetAddress that is part of the transaction. So the TargetAddress is made in reverse, we look into the TxOut and extract an address, and convert it to TargetAddress.

A better approach, is to 
- generate addresses
- generate events based on those addresses
- randomly select a an address as TargetAddress and run the test.
-
This required a bit more work, so I opted out to do the work in the [PLT-1360](https://input-output.atlassian.net/browse/PLT-1360)